### PR TITLE
Add OnRewardRedeemed, OnCustomRewardCreated, OnCustomRewardUpdated, OnLeaderboardSubs & OnLeaderboardBits event

### PR DESCRIPTION
**Pull request by JayJay1989**
After some reverse engineering i added following functionality:
- OnRewardRedeemed event
 This will be fired when someone redeems a reward.
- OnCustomRewardUpdated
 This will be fired when a custom reward has been changed.

Note: Highlighted message are part of the irc messages.

### CODE REVIEW

**Comment by swiftyspiffy** [comment]
This looks great. Do you have sample payloads received from Twitch for these events? Thanks!

**Comment by JayJay1989** [comment]
Sure, here you go: https://gist.github.com/JayJay1989/1730d1ede2e2c23809c6986af93e0c14

**Comment by JayJay1989** [comment]
i added the payloads for the leaderboard events for giftsubs and bits also to the link above

**Comment by swiftyspiffy** [comment]
Looks good to me. Will leave open a few more days to allow others to review before merging. Thanks!

**Comment by JayJay1989** [comment]
@swiftyspiffy the events doesn't get triggered for some reason. Maybe its something i do wrong in my test setup.

**Comment by swiftyspiffy** [comment]
So it is mergeable then?

**Comment by JayJay1989** [comment]
Channelpoint reward redeemed should working: http://prntscr.com/qbsbml
Leaderboard events should working now: http://prntscr.com/qbsp9s

**Comment by wtfblub** [comment]
Twitch added a official PubSub API for reward redeems yesterday: https://dev.twitch.tv/docs/pubsub
Looks like only the topic name is different

**Comment by JayJay1989** [comment]
Reason why i implemented 'community-points-channel-v1' and not 'channel-points-channel-v1'

community-points-channel-v1:

-  it can trigger for other channels than yourself. Ideal for twitch integration services.

channel-points-channel-v1:

- only triggers for your own channel

File: TwitchLib.PubSub/Models/Responses/Messages/CommunityPointsChannel.cs [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+            switch (Type)
+            {
+                case CommunityPointsChannelType.RewardRedeemed:
+                    ChannelId = json.SelectToken("data.redemption.channel_id").ToString();
+                    Login = json.SelectToken("data.redemption.user.login").ToString();
+                    DisplayName = json.SelectToken("data.redemption.user.display_name").ToString();
+                    RewardTitle = json.SelectToken("data.redemption.reward.title").ToString();
+                    RewardPrompt = json.SelectToken("data.redemption.reward.prompt").ToString();
+                    RewardCost = int.Parse(json.SelectToken("data.redemption.reward.cost").ToString());
+                    Message = json.SelectToken("data.user_input")?.ToString();
```
Comment (wtfblub) [inline TwitchLib.PubSub/Models/Responses/Messages/CommunityPointsChannel.cs]: This changed to `data.redemption.user_input` last night

Comment (JayJay1989) [reply TwitchLib.PubSub/Models/Responses/Messages/CommunityPointsChannel.cs]: > This changed to `data.redemption.user_input` last night

Also on the community-points-channel-v1 ?

File: TwitchLib.PubSub/Models/Responses/Message.cs [hunk 2]
```diff
@@ -50,6 +50,12 @@ public Message(string jsonStr)
                 case "following":
                     MessageData = new Following(encodedJsonMessage);
                     break;
+                case "community-points-channel-v1":
+                    MessageData = new CommunityPointsChannel(encodedJsonMessage);
+                    break;
+                case "leaderboard-events-v1":
+                    MessageData = new LeaderboardEvents(encodedJsonMessage);
+                    break;
```
Comment (swiftyspiffy) [inline TwitchLib.PubSub/Models/Responses/Message.cs]: These are presumably undocumented (I didn't see them in the docs). We might want to note that somewhere going forward.

File: TwitchLib.PubSub/TwitchLib.PubSub.csproj [hunk 3]
```diff
@@ -22,8 +22,8 @@
     <FileVersion>3.1.3</FileVersion>
   </PropertyGroup>
   <ItemGroup>
-    <PackageReference Include="Newtonsoft.Json" Version="12.0.1" />
-    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.2.0" />
+    <PackageReference Include="Newtonsoft.Json" Version="11.0.2" />
+    <PackageReference Include="Microsoft.Extensions.Logging.Abstractions" Version="2.1.1" />
```
Comment (swiftyspiffy) [inline TwitchLib.PubSub/TwitchLib.PubSub.csproj]: I'm not sure we want to make these changes. Is there a good reason for this?

File: TwitchLib.PubSub/Models/Responses/Message.cs (hunk 2 above)
Comment (JayJay1989) [reply TwitchLib.PubSub/Models/Responses/Message.cs]: Where can i add some notes to this feature? Its a weekly leaderboard for subgifts and bits.

**Comment by swiftyspiffy** [comment]
Got confirmation this code has been running in a prod environment for several days.

**Comment by Pwnoz0r** [comment]
Anyway you can provide your example of this working?

I've cloned the latest dev branch (https://github.com/TwitchLib/TwitchLib.PubSub/commit/1b4f59ef46245f31965bc3772c1ac5ee2c32b066) and I'm unable to get the events to fire.

Snippet of what isn't working:

```
PubSub = new TwitchPubSub(Logger);
PubSub.OnPubSubServiceConnected += RewardHelper.PubSub_OnPubSubServiceConnected;
PubSub.OnListenResponse += RewardHelper.PubSub_OnListenResponse;
PubSub.OnRewardRedeemed += RewardHelper.PubSub_OnRewardRedeemed;
PubSub.Connect();
```
`PubSub_OnPubSubServiceConnected`
```
Startup.PubSub.ListenToRewards("<channel id>");
Startup.PubSub.SendTopics("auth token with channel:read:redemptions set");
```

OnPubSubServiceConnected and OnPubSubServiceConnected both fire stating the topics were successfully subscribed to. I've tested it without the auth token and it fails to subscribe to the topic. I've set breakpoints where the message is received / parsed and that breakpoint was never hit.

**Comment by JayJay1989** [comment]
Here you go: https://github.com/JayJay1989/TwitchLib.ExampleBotOp 26 jan. 2020 05:52 schreef Jonathan Rainier <notifications@github.com>:Anyway you can provide your example of this working?
I've cloned the latest dev branch (1b4f59e) and I'm unable to get the events to fire.
Snippet of what isn't working:
PubSub = new TwitchPubSub(Logger);
PubSub.OnPubSubServiceConnected += RewardHelper.PubSub_OnPubSubServiceConnected;
PubSub.OnListenResponse += RewardHelper.PubSub_OnListenResponse;
PubSub.OnRewardRedeemed += RewardHelper.PubSub_OnRewardRedeemed;
PubSub.Connect();

PubSub_OnPubSubServiceConnected
Startup.PubSub.ListenToRewards("<channel id>");
Startup.PubSub.SendTopics("auth token with channel:read:redemptions set");

OnPubSubServiceConnected and OnPubSubServiceConnected both fire stating the topics were successfully subscribed to. I've tested it without the auth token and it fails to subscribe to the topic. I've set breakpoints where the message is received / parsed and that breakpoint was never hit.

**Comment by Pwnoz0r** [comment]
> Here you go: https://github.com/JayJay1989/TwitchLib.ExampleBotOp 26 jan. 2020 05:52 schreef Jonathan Rainier <notifications@github.com>:Anyway you can provide your example of this working? I've cloned the latest dev branch ([1b4f59e](https://github.com/TwitchLib/TwitchLib.PubSub/commit/1b4f59ef46245f31965bc3772c1ac5ee2c32b066)) and I'm unable to get the events to fire. Snippet of what isn't working: PubSub = new TwitchPubSub(Logger); PubSub.OnPubSubServiceConnected += RewardHelper.PubSub_OnPubSubServiceConnected; PubSub.OnListenResponse += RewardHelper.PubSub_OnListenResponse; PubSub.OnRewardRedeemed += RewardHelper.PubSub_OnRewardRedeemed; PubSub.Connect(); PubSub_OnPubSubServiceConnected Startup.PubSub.ListenToRewards("<channel id>"); Startup.PubSub.SendTopics("auth token with channel:read:redemptions set"); OnPubSubServiceConnected and OnPubSubServiceConnected both fire stating the topics were successfully subscribed to. I've tested it without the auth token and it fails to subscribe to the topic. I've set breakpoints where the message is received / parsed and that breakpoint was never hit. —You are receiving this because you authored the thread.Reply to this email directly, view it on GitHub, or unsubscribe.

Thank you for the prompt reply! Mine appears to be working now after no changes to my logic, I ran both side by side for a couple of days and all events are now firing.

**Pull request merged**